### ISSUE TITLE
Enhance supported fill value type for constant pad in functional.pad

### ISSUE THREAD

**Issue by voldemortX** [pytorch/pytorch#50566]
## 🚀 Feature
Support fill value as list/tuple (Sequence) for constant padding in `nn.functional.pad()`.

## Motivation

In torchvision we need to support fill value as list/tuple (Sequence) for constant padding an Tensor image where we rely on torch's pad function. 
https://github.com/pytorch/vision/issues/3227
https://github.com/pytorch/vision/issues/3224

## Alternatives

We can do some workaround to support it in torchvision, but it would be highly inefficient.

I can try implement this myself if somebody can give me some guidance on how a feature like this is implemented. Since I've not contributed to this repo before.

cc @datumbox @mruberry

**Comment by mruberry** [comment]
Thank you for this suggestion, @voldemortX. This feature makes sense and we would accept a PR implementing it.

Would an implementation of [NumPy's pad function](https://numpy.org/doc/stable/reference/generated/numpy.pad.html?highlight=pad#numpy.pad) in PyTorch solve this issue, too?

**Comment by voldemortX** [comment]
> Would an implementation of [NumPy's pad function](https://numpy.org/doc/stable/reference/generated/numpy.pad.html?highlight=pad#numpy.pad) in PyTorch solve this issue, too?

@mruberry Do you mean the numpy version of pad should be implemented?

The kind of padding we use in vision tasks is to pad only the channel axis, for instance, 3 different values for R,G,B channels respectively. Hence the need for a list/tuple input for each channel. I think the numpy pad only support before/after value for each axis:

```
x = np.ones((3,2,2))  # A 2x2 RGB image
np.pad(x, ((0,), (1,), (1, )), 'constant', constant_values=((0,), (2, ), (3, )))
array([[[3., 2., 2., 3.],
        [3., 1., 1., 3.],
        [3., 1., 1., 3.],
        [3., 2., 2., 3.]],
       [[3., 2., 2., 3.],
        [3., 1., 1., 3.],
        [3., 1., 1., 3.],
        [3., 2., 2., 3.]],
       [[3., 2., 2., 3.],
        [3., 1., 1., 3.],
        [3., 1., 1., 3.],
        [3., 2., 2., 3.]]])
```

While we want a feature like this:

```
pad(x, (1,), fill=(0,2,3))
array([[[0., 0., 0., 0.],
        [0., 1., 1., 0.],
        [0., 1., 1., 0.],
        [0., 0., 0., 0.]],
       [[2., 2., 2., 2.],
        [2., 1., 1., 2.],
        [2., 1., 1., 2.],
        [2., 2., 2., 2.]],
       [[3., 3., 3., 3.],
        [3., 1., 1., 3.],
        [3., 1., 1., 3.],
        [3., 3., 3., 3.]]])
```

**Comment by mruberry** [comment]
> @mruberry Do you mean the numpy version of pad should be implemented?

Yes, we do want to implement np.pad in PyTorch, but it doesn't have to be the solution to this issue. I'm was just curious if implementing it would be a solution. Thank you for explaining the difference between it and what you're looking for.
 
> EDIT: I just realized that it seems the fill parameter could be as large as the original tensor in this way (considering not only the channel axis)... Do you think it is possible to implement?

I'm not totally sure what you mean, sorry. Could you give an example of what you're thinking?

**Comment by voldemortX** [comment]
> EDIT: I just realized that it seems the fill parameter could be as large as the original tensor in this way (considering not only the channel axis)... Do you think it is possible to implement?

@mruberry Sorry. I'm not sure what I mean either... Let's ignore that statement.

But I do think the list/tuple fill could be quite tricky if we consider it can be a list for every axis?

**Comment by mruberry** [comment]
It's hard to say without thinking about it more thoroughly, but yes it would probably be tricky to implement. It may even require specialized parsing.

**Comment by voldemortX** [comment]
@mruberry What about only allowing list fill for the channel axis, and implement it as a new padding mode (for instance, image_constant)? Since in image processing, we try to match the functional of tensors with image libraries such as [PIL's pad](https://pillow.readthedocs.io/en/stable/reference/ImageOps.html?highlight=pad#PIL.ImageOps.pad), only padding on the channel axis.

**Comment by mruberry** [comment]
Adding a padding mode to support this behavior, especially if it's consistent with an existing library's padding functionality, sounds like a great way to address this issue, @voldemortX.

**Comment by voldemortX** [comment]
Great, I'll try it out in a PR in a few days.
For now, I'd like to go with an assumption of input tensors with at least 3 dimensions and define the 3rd to last as the 'color' channel, i.e. [..., C, d1, d2]. Can you suggest how I should proceed, should I create a new cpp file in https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native and modify `torch.nn.functional.pad()` ? Afraid I'm not quite familiar with the code structure yet...

**Comment by mruberry** [comment]
pad's Python exposure and doc string are here: 

https://github.com/pytorch/pytorch/blob/48b6b9221a544f7d5b7d6f93eec2fa304da8af94/torch/nn/functional.py#L3941

You can follow the approach that circular padding takes where the implementation is exclusively in Python:

https://github.com/pytorch/pytorch/blob/48b6b9221a544f7d5b7d6f93eec2fa304da8af94/torch/nn/functional.py#L4353

And not available from the C++ API, or the approach that other padding modes take where the Python entrypoint calls a C++ version of the operation.

Before we get there, however, let's talk a little bit more about the design. What's the closest existing function from a popular library to the behavior you're proposing that we should use as a reference?

**Comment by voldemortX** [comment]
That would be PIL (https://pillow.readthedocs.io/en/stable/reference/ImageOps.html?highlight=pad#PIL.ImageOps.pad), since that is what we are trying to match (exceed) in torchvision. By exceed, I'd like it to process arbitrary leading dimensions and not limiting the size of 3rd to last dimension:
https://github.com/pytorch/vision/blob/f16322b596c7dc9e9d67d3b40907694f29e16357/torchvision/transforms/functional.py#L384

**Comment by mruberry** [comment]
Doesn't the Torchvision op support an arbitrary number of leading dimensions? It's documented as doing so:

https://github.com/pytorch/vision/blob/f16322b596c7dc9e9d67d3b40907694f29e16357/torchvision/transforms/functional.py#L387

Can you elaborate on what you mean by "not limiting the size of 3rd to last dimension?" I'm not sure what you mean. Maybe I just need it explained in a different way.

**Comment by voldemortX** [comment]
By the 3rd dimension I mean PIL images have only a few modes (RGB - HWx3, RGBA HWx4, etc.) but for tensors we can have naturally more freedom HWxN.  The current torchvision tensor ops mostly all support the arbitrary leading dimension and 3rd dimension thing, I just meant to not lose that property for this pad feature.

**Comment by mruberry** [comment]
Got it. I think this makes sense. Let's have @fmassa weigh-in, too. If he likes the idea we can proceed to look at a PR.

**Comment by sidijju** [comment]
Was this issue ever resolved? I could also try implementing a fix for this if I could get some guidance?

I've run into a use case where I'd like to pad an image based on the mean pixel value for each channel (R, G, B), and it seems like a successful implementation of this feature would resolve that nicely. Currently it seems like this isn't supported for tensors.

**Comment by zhazhuanling12** [comment]
I'm curious about the current status of this issue. Is the Half data type already supported?

**Comment by voldemortX** [comment]
@zhazhuanling12 @sidijju To my recollections, I did not actually proceed with this pr. You can check the current pytorch doc to see if others have done this.
